File: quickstart_guides/recursion/java/UniqueBST.java
/*
 * Title: Unique BST
 *
 * Problem:
 *  Construct all possible BSTs for keys 1 to N.
 *
 * Execution: javac UniqueBST.java && java UniqueBST
 *
 */

import java.util.ArrayList;

class Node
{
    int key;
    Node left, right;
    Node(int data)
    {
        this.key=data;
        left=right=null;
    }
}

public class UniqueBST {

    static ArrayList<Node> constructTrees(int start, int end) {
        ArrayList<Node> list=new ArrayList<>();

        // If start > end then subtree will be empty so returning NULL in list.
        if (start > end)
        {
            list.add(null);
            return list;
        }

        // Iterate through all values for constructing left and right.
        for (int i = start; i <= end; i++) {
            // Constructing left subtree.
            ArrayList<Node> leftSubtree  = constructTrees(start, i - 1);

            // Constructing right subtree.
            ArrayList<Node> rightSubtree = constructTrees(i + 1, end);

            // Loop through all left and right subtrees and connecting
            // them to ith root  below  */
            for (int j = 0; j < leftSubtree.size(); j++) {
                Node left = leftSubtree.get(j);
                for (int k = 0; k < rightSubtree.size(); k++) {
                    Node right = rightSubtree.get(k);
                    Node node = new Node(i);        // making value i as root
                    node.left = left;              // connect left subtree
                    node.right = right;            // connect right subtree
                    list.add(node);                // add this tree to list
                }
            }
        }
        return list;
    }

    // A utility function to do pre-order traversal of BST
    static void preorder(Node root) {
        if (root != null)
        {
            System.out.print(root.key+" ") ;
            preorder(root.left);
            preorder(root.right);
        }
    }

    public static void main(String args[]) {
        ArrayList<Node> totalTreesFrom1toN = constructTrees(1, 3);

        System.out.println("Preorder traversals of all constructed BSTs are ");
        for (int i = 0; i < totalTreesFrom1toN.size(); i++) {
            preorder(totalTreesFrom1toN.get(i));
            System.out.println();
        }
    }
}

File: quickstart_guides/recursion/python/unique_bst.py
"""
Title: Unique BST.

Problem:
    Construct all possible BSTs for keys 1 to N.

Execution: python unique_bst.py
"""
from typing import Any, List
import unittest


class Node:
    def __init__(self, item):
        self.key = item
        self.left = None
        self.right = None


def preorder(root, lst: List[int]) -> List[int]:
    if root:
        lst.append(root.key)
        preorder(root.left, lst)
        preorder(root.right, lst)
    return lst


def construct_trees(start: int, end: int) -> List[Any]:
    node_list = []

    # If start > end, then subtree will be empty, so return None in the list.
    if start > end:
        node_list.append(None)
        return node_list

    # Iterate through all values from start to end for constructing left and
    # right subtree.
    for i in range(start, end + 1):

        # Construct left subtree.
        left_subtree = construct_trees(start, i - 1)

        # Construct right subtree.
        right_subtree = construct_trees(i + 1, end)

        # Loop through all left and right subtrees.
        for j in range(len(left_subtree)):
            left = left_subtree[j]
            for k in range(len(right_subtree)):
                right = right_subtree[k]
                node = Node(i)
                node.left = left
                node.right = right
                node_list.append(node)
    return node_list


class TestUniqueBST(unittest.TestCase):
    """Unit tests for unique_bst."""

    def test_1(self):
        expected_out = [[1, 2, 3], [1, 3, 2], [2, 1, 3], [3, 1, 2], [3, 2, 1]]

        total_trees = construct_trees(1, 3)

        for i in range(len(total_trees)):
            res = preorder(total_trees[i], [])
            self.assertEqual(res, expected_out[i])


if __name__ == "__main__":
    unittest.main()
